nrf_security: Doc updates for mbed TLS 3.0.0 (#652)
nrfconnect/sdk-nrfxlib: nrf_security/doc/drivers.rst

The :ref:`nrf_oberon_readme` is distributed as a closed-source binary that provides select cryptographic algorithms optimized for use in nRF devices.
This provides faster execution than the original Mbed TLS implementation.

The nrf_oberon driver provides support for AES ciphers, SHA-1, SHA-256, SHA-384, SHA-512, ECDH and ECDSA using NIST curve secp224r1 and secp256r1, and ECJPAKE using NIST curve secp256r1.

Enabling the nrf_oberon driver
==============================

The :ref:`nrf_oberon_readme` PSA driver can be enabled by setting the :kconfig:`PSA_CRYPTO_DRIVER_OBERON` Kconfig variable.

.. _nrf_security_drivers_builtin:

Builtin Mbed TLS
****************

Builtin Mbed TLS uses the open-source software from the `Arm Mbed TLS project`_ with minimal adjustments in PSA Core and PSA driver scope to support multiple drivers.

Builtin Mbed TLS support can be used to add support for features not available in the `Arm CryptoCell cc3xx driver`_, for example AES-256 or ECC Brainpool curve types.

Similarly, you can use the builtin Mbed TLS to add support for features not available in the `nrf_oberon driver`_, for example to add support for more ECC curve types.

.. note:

>    nrf_security provides a replacement to the file ``psa_crypto_driver_wrappers.c`` with adjustments to support multiple drivers.

Reviewer: This sounds like we have added a file.

But we have patched the file in mbedtls, right?

Author: Currently we have a copy of this file, it is not in our mbedtls fork

.. _nrf_security_drivers:

nrf_security drivers
####################

.. contents::
   :local:
   :depth: 2

> The nrf_security module supports multiple enabled PSA drivers at the same time.

Author: Will be handled in an upcoming improvement task

This mechanism is intended to extend the available feature set of hardware-accelerated cryptography or to provide alternative implementations of the PSA Crypto APIs.

Enabling a cryptographic feature or algorithm is done using PSA Crypto API configurations that follows the format ``PSA_WANT_ALG_XXXX``.

Enabling more than one PSA driver may add support for additional key sizes or modes of operation. 

It is possible to disable specific features on PSA driver level to optimize the code size.

The nrf_security module supports the following PSA drivers:

* Arm CryptoCell cc3xx
* nrf_oberon binary library

.. note::

>    Note that whenever this documentation mentions 'original' Mbed TLS, it refers to the open-source `Arm Mbed TLS project`_, not the customized version available in Zephyr.

Reviewer: I thought that when this documentation metions 'origina' Mbed TLS it refers to a nordic-customized version of the customized version available in Zephyr ?

Reviewer: Note that we have forked Zephyr's fork.

Author: We need to update to give info about our fork. This was only true earlier... Will be fixed

Author: Fixed

   There is an option to utilize a 'builtin' driver, which corresponds to software implemented cryptography from the 'original' Mbed TLS deliverables. 
   This is provided to ensure that the cryptographic toolbox supports all requested features.

.. _nrf_security_drivers_cc3xx:

Arm CryptoCell cc3xx driver
***************************

> The Arm CryptoCell cc3xx driver is a runtime library that provides hardware-accelerated cryptography using the Arm CryptoCell cc310/cc312 hardware.

Reviewer: The interesting part is that this library is not delivered in source form.

So

"is a closed-source binary library that provides"

Author: Fixed (part of next force-push)

Reviewer: let's use the same phrasing as oberon

The Arm CryptoCell cc3xx driver is only available on the following devices:

* nRF52840
* nRF9160
* nRF5340

Enabling the Arm CryptoCell cc3xx driver
========================================

The Arm CryptoCell cc3xx driver can be enabled by setting the :kconfig:`PSA_CRYPTO_DRIVER_CC3XX` Kconfig variable.

Using the Arm CryptoCell cc3xx driver
=====================================

To use the :ref:`nrf_cc3xx_mbedcrypto_readme` PSA driver, the Arm CryptoCell cc310/cc312 hardware must be first initialized.

The Arm CryptoCell cc3xx hardware is initialized in :file:`<NCS>/nrf/drivers/hw_cc310/hw_cc310.c` and is controlled with the :kconfig:`CONFIG_HW_CC3XX` Kconfig variable.
The Kconfig variable has a default value of 'y' when cc3xx is available in the SoC.

.. _nrf_security_driver_oberon:

nrf_oberon driver
*****************

> The :ref:`nrf_oberon_readme` is distributed as a closed-source binary that provides select cryptographic algorithms optimized for use in nRF devices.

Reviewer: ```suggestion
The :ref:`nrf_oberon_readme` is distributed as a closed-source binary library that provides select cryptographic algorithms optimized for use in nRF devices.
```

Or maybe omit library, but we should describe the two drivers in the same way at least as they are both binary blobs.

Author: Fixed, part of next force-push. Removed library and put emphasis on "binary" wherever mentioned

This provides faster execution than the original Mbed TLS implementation.

The nrf_oberon driver provides support for AES ciphers, SHA-1, SHA-256, SHA-384, SHA-512, ECDH and ECDSA using NIST curve secp224r1 and secp256r1, and ECJPAKE using NIST curve secp256r1.

Enabling the nrf_oberon driver
==============================

> To use the :ref:`nrf_oberon_readme` as a backend, set the :kconfig:`PSA_CRYPTO_DRIVER_OBERON` Kconfig variable to true.

Reviewer: Use the same sentence as for cryptocell.

Author: Fixed, part of next force-push

We are not using the term "backend" for drivers so the file needs a bit more cleanup

.. _nrf_security_backends_orig_mbedtls:

Builtin Mbed TLS
****************

Builtin Mbed TLS uses the open-source software from the `Arm Mbed TLS project`_.

> The software is unaltered from the Arm Mbed TLS project source distribution, meaning it is in its `vanilla` form.

Reviewer: No it's not.

Reviewer: vanilla, builtin, original,

we need to stick to one word.

Author: Vanilla can be removed, as this was only available when we had vanilla as an option for configurations.
Fixed, part of the next force-push

> You can use the builtin Mbed TLS support to add support for features not available in the `Arm CryptoCell cc3xx backend`_, for example AES-256 or ECC Brainpool curve types.

Reviewer: I can use?

nrf_security uses the builtin Mbed TLS support to add support for features not available ...

Author: Fixed, using more technical phrasing

Similarly, you can use the builtin Mbed TLS to add support for features not available in the `nrf_oberon backend`_, for example to support more ECC curve types.

.. note:

>    Certain changes has been made tp the file ``pss_crypto_driver_wrappers`` to provide support for the PSA drivers.

Reviewer: ```suggestion
   Changes have been made to the Mbed TLS source file ``pss_crypto_driver_wrappers.c`` to provide support for the PSA drivers.
```

Enabling the builtin Mbed TLS support
=====================================

> To enable the builtin Mbed TLS support, set the :kconfig:`PSA_CRYPTO_DRIVER_BUILTIN` Kconfig variable to true.

Reviewer: Use the same sentence as cryptocell and oberon

Author: Fixed, part of next force-push

It is generally not needed to enable builtin Mbed TLS support manually, as there is Kconfig logic that does that, depending on the enabled cryptographic features or algorithms and the configuration of PSA drivers.

Legacy Mbed TLS
***************

> Some legacy Mbed TLS APIs are still supported, for instance for TLS and DTLS support and backwards compatibility in Zephyr subsystems.

Reviewer: ```suggestion
Some legacy Mbed TLS APIs are still supported, for instance for TLS and DTLS support and for backwards compatibility.
```

It's not just zephyr subsystems using mbedtls legacy. It's customer code as well.

Author: True, fixed, part of next force-push

> Some legacy Mbed TLS APIs are still supported, for instance for TLS and DTLS support and backwards compatibility.

Author: Usage of these APIs will probably be too wide to explain. You can use it in bootloader, provisioning, protocol, just to name a few use-cases where the legacy APIs can be used.

Enabling legacy APIs requires enabling one of the available PSA drivers.

.. note::

>    * The legacy Mbed TLS APIs no longer support glued functionality.

Author: I don't think that adding "the" in front of this changes much so I accept the addition

Fixed, part of next force-push

>    Legacy configurations has no effect on the configurations for the secure image of a TF-M build.

Reviewer: ```suggestion
   Legacy configurations also no longer have an effect on the configurations for the secure image of a TF-M build.
```

Author: I'm changing the notes to bullet-points, so the also is superfluous

Author: Changed line 104 as well, as it looked a bit strange

Reviewer: You removed the part about "no longer", which is important as it is a backwards incompatible change.

Author: Fixed

Enabling legacy Mbed TLS support
================================

> The legacy Mbed TLS APIs can be configured by setting the option kconfig:`NORDIC_SECURITY_BACKEND` instead of setting the option kconfig:`NRF_SECURITY`.

Reviewer 2: ```suggestion
You can configure the legacy Mbed TLS APIs by setting the :kconfig:`NORDIC_SECURITY_BACKEND` Kconfig options instead of setting :kconfig:`NRF_SECURITY`.
```
`CONFIG_` before both names?

Author: Outdated proposal

> Additionally the the options :kconfig:`CC3XX_BACKEND`or :kconfig:`OBERON_BACKEND` must be set, to enable the desired backend support.

Reviewer: ```suggestion
Additionally either  :kconfig:`CC3XX_BACKEND` or :kconfig:`OBERON_BACKEND` must be enabled.
```

Author: Fixed, part of next force-push

> Additionally either :kconfig:`CC3XX_BACKEND` or :kconfig:`OBERON_BACKEND` must be enabled.

Reviewer 2: ```suggestion
Additionally, you must enable either :kconfig:`CONFIG_CC3XX_BACKEND` or :kconfig:`CONFIG_OBERON_BACKEND` Kconfig option.
```
I'd make this sentence part of the previous paragraph, as this is also about configuration.

Author: Outdated proposal

.. note::

>    Enabling the CryptoCell backend non-secure image of a TF-M build will have no effect.

Reviewer: ```suggestion
   Enabling the CryptoCell backend in a non-secure image of a TF-M build will have no effect.
```

Reviewer: What about enabling cryptocell backend in a non-secure image for a non-TF-M build ?

Author: Changed to **Enabling the CryptoCell by using :kconfig:`CC3XX_BACKEND` in a non-secure image of a TF-M build will have no effect.** to make it absolutely clear that we are talking about a legacy (backend is only usable in scope of Legacy APIs and configuration)

>    Enabling the CryptoCell by using :kconfig:`CC3XX_BACKEND` in a non-secure image of a TF-M build will have no effect.

Reviewer 2: ```suggestion
   Enabling the CryptoCell by using :kconfig:`CONFIG_CC3XX_BACKEND` in a non-secure image of a TF-M build will have no effect.
```

Author: Fixed, part of the next force-push
